### PR TITLE
CSPL-947: Added admin password verification on splunk instance

### PR DESCRIPTION
**Pull request by tpereirasplunk**
# Changes

- Added admin password verification to secret object update and delete cases
- Split Licensemaster test case's into individual files based on SVA

Note: Hec_token would also be verified as part of secret object update api access key. Will be added later

**Test Scenarios**: (SVA's covered: S1)
Delete secret data
Verify New versioned secret are created with correct value.
Verify new secrets are mounted on pods.
Verify old Secrets cannot be used for api access (password)

**Test Scenarios**: (SVA's covered: S1, C3, M4)
Update Secrets Data
Verify New versioned secret are created with correct value.
Verify new secrets are mounted on pods.
Verify New Secrets are present in server.conf 
Verify New updated Secrets via api access (password)

# Passing Test Runs
**Secret**
C3/M4 Modify
```
[2] • [SLOW TEST:1906.794 seconds]
[2] Secret Test for M4 SVA
[2] /Users/tpereira/Git/splunk-operator/test/secret/secret_m4_test.go:26
[2]   Multisite cluster deployment (M13 - Multisite indexer cluster, Search head cluster)
[2]   /Users/tpereira/Git/splunk-operator/test/secret/secret_m4_test.go:46
[2]     secret, integration: secret update on multisite indexers and search head cluster
[2]     /Users/tpereira/Git/splunk-operator/test/secret/secret_m4_test.go:47
[2] ------------------------------
[2] {"level":"info","ts":1617130451.0888782,"msg":"testenv deleted.\n","testenv":"secret-ut0"}
[2]
[2] JUnit report was created: /Users/tpereira/Git/splunk-operator/test/secret/secret-ut0_junit.xml
[2]
[2] Ran 1 of 5 Specs in 1913.612 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 4 Pending | 0 Skipped
[2] PASS

[1] • [SLOW TEST:1839.767 seconds]
[1] Secret Test for SVA C3
[1] /Users/tpereira/Git/splunk-operator/test/secret/secret_c3_test.go:26
[1]   Clustered deployment (C3 - clustered indexer, search head cluster)
[1]   /Users/tpereira/Git/splunk-operator/test/secret/secret_c3_test.go:46
[1]     secret: secret update on indexers and search head cluster
[1]     /Users/tpereira/Git/splunk-operator/test/secret/secret_c3_test.go:47
[1] ------------------------------
[1] {"level":"info","ts":1617128408.768227,"msg":"testenv deleted.\n","testenv":"secret-ur6"}
[1]
[1] JUnit report was created: /Users/tpereira/Git/splunk-operator/test/secret/secret-ur6_junit.xml
[1]
[1] Ran 1 of 3 Specs in 1846.601 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 2 Pending | 0 Skipped
[1] PASS
```
S1 Modify/delete
```
[1] • [SLOW TEST:556.052 seconds]
[1] Secret Test for SVA S1
[1] /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:26
[1]   Standalone deployment (S1) with LM
[1]   /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:46
[1]     secret: Secret update on a standalone instance
[1]     /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:47
[1] ------------------------------
[1] {"level":"info","ts":1617124303.36332,"msg":"testenv deleted.\n","testenv":"secret-uyy"}
[1]
[1] JUnit report was created: /Users/tpereira/Git/splunk-operator/test/secret/secret-uyy_junit.xml
[1]
[1] Ran 1 of 1 Specs in 562.926 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[1] PASS

[1] • [SLOW TEST:531.297 seconds]
[1] Secret Test for SVA S1
[1] /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:26
[1]   Standalone deployment (S1) with LM
[1]   /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:122
[1]     secret: Secret Object is recreated on delete and new secrets are applied to Splunk Pods
[1]     /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:123
[1] ------------------------------
[1] {"level":"info","ts":1617125091.854136,"msg":"testenv deleted.\n","testenv":"secret-dc4"}
[1]
[1] JUnit report was created: /Users/tpereira/Git/splunk-operator/test/secret/secret-dc4_junit.xml
[1]
[1] Ran 1 of 1 Specs in 538.122 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[1] PASS

[2] • [SLOW TEST:338.035 seconds]
[2] Secret Test for SVA S1
[2] /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:26
[2]   Standalone deployment (S1)
[2]   /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:194
[2]     secret, smoke, integration: Secret Object data is repopulated in secret object on passing empty Data map and new secrets are applied to Splunk Pods
[2]     /Users/tpereira/Git/splunk-operator/test/secret/secret_s1_test.go:195
[2] ------------------------------
[2] {"level":"info","ts":1617125777.108496,"msg":"testenv deleted.\n","testenv":"secret-bg3"}
[2]
[2] JUnit report was created: /Users/tpereira/Git/splunk-operator/test/secret/secret-bg3_junit.xml
[2]
[2] Ran 1 of 3 Specs in 344.830 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 2 Pending | 0 Skipped
[2] PASS
```
**Licensemaster**
```
[2] • [SLOW TEST:842.198 seconds]
[2] Licensemaster test
[2] /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_m4_test.go:25
[2]   Multisite cluster deployment (M4 - Multisite indexer cluster, Search head cluster)
[2]   /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_m4_test.go:45
[2]     licensemaster: Splunk Operator can configure license master with indexers and search head in M4 SVA
[2]     /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_m4_test.go:46
[2] ------------------------------
[2] {"level":"info","ts":1617138692.6694698,"msg":"testenv deleted.\n","testenv":"licensemaster-wp"}
[2]
[2] JUnit report was created: /Users/tpereira/Git/splunk-operator/test/licensemaster/licensemaster-wp_junit.xml
[2]
[2] Ran 1 of 3 Specs in 849.008 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 2 Pending | 0 Skipped
[2] PASS

2] • [SLOW TEST:625.027 seconds]
[2] Licensemaster test
[2] /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_c3_test.go:25
[2]   Clustered deployment (C3 - clustered indexer, search head cluster)
[2]   /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_c3_test.go:45
[2]     licensemaster, integration: Splunk Operator can configure License Master with Indexers and Search Heads in C3 SVA
[2]     /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_c3_test.go:46
[2] ------------------------------
[2] {"level":"info","ts":1617137701.714098,"msg":"testenv deleted.\n","testenv":"licensemaster-bt"}
[2]
[2] JUnit report was created: /Users/tpereira/Git/splunk-operator/test/licensemaster/licensemaster-bt_junit.xml
[2]
[2] Ran 1 of 1 Specs in 632.037 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[2] PASS

[2] • [SLOW TEST:316.097 seconds]
[2] Licensemaster test
[2] /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_s1_test.go:25
[2]   Standalone deployment (S1) with LM
[2]   /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_s1_test.go:45
[2]     licensemaster: Splunk Operator can configure License Master with Standalone in S1 SVA
[2]     /Users/tpereira/Git/splunk-operator/test/licensemaster/lm_s1_test.go:46
[2] ------------------------------
[2] {"level":"info","ts":1617137044.6949341,"msg":"testenv deleted.\n","testenv":"licensemaster-ah"}
[2]
[2] JUnit report was created: /Users/tpereira/Git/splunk-operator/test/licensemaster/licensemaster-ah_junit.xml
[2]
[2] Ran 1 of 1 Specs in 322.963 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[2] PASS
```

### CODE REVIEW

File: test/testenv/secretutil.go [hunk 1]
```diff
@@ -159,3 +159,27 @@ func GetSecretDataMap(hecToken string, password string, pass4SymmKey string, idx
 	}
 	return updatedSecretData
 }
+
+// CheckSecretViaAPI return api response if it can be accessed
```
Comment (pdhanoyasplunk) [inline test/testenv/secretutil.go:163]: The function is not performing a check. Maybe its better to call it `ExecuteCommandInPod` we can then use it in other places. 
We can return `(string, error)` and leave error handling part to caller.

Comment (pdhanoyasplunk) [inline test/testenv/secretutil.go:163]: The function is not performing a check. Maybe it's better to call it `ExecuteCommandInPod` we can then use it in other places. 
We can return `(string, error)` and leave error handling part to caller.

File: test/testenv/verificationutils.go [hunk 2]
```diff
@@ -489,3 +510,18 @@ func VerifySplunkServerConfSecrets(deployment *Deployment, testenvInstance *Test
 		}
 	}
 }
+
+// VerifySplunkSecretAPIKeys check if keys can be used to access api i.e validate they are authentic
```
Comment (pdhanoyasplunk) [inline test/testenv/verificationutils.go:514]: `VerifySplunkSecretAPIKeys` vs `VerifySplunkSecretViaAPI`

**Review by pdhanoyasplunk — CHANGES_REQUESTED** [review]

File: test/testenv/secretutil.go (hunk 1 above)
Comment (tpereirasplunk) [reply test/testenv/secretutil.go]: Good idea. Modified and moved to util.

File: test/testenv/util.go [hunk 3]
```diff
@@ -521,3 +521,15 @@ func GetConfLineFromPod(podName string, filePath string, ns string, configName s
 	}
 	return config, err
 }
+
+// ExecuteCommandOnPod execute command on given pod and return result
+func ExecuteCommandOnPod(deployment *Deployment, podName string, stdin string) (string, error) {
```
Comment (akondur) [inline test/testenv/util.go:526]: Can't we use PodExecCommand directly? Any reason why we are creating a function?

**Comment by pdhanoyasplunk** [comment]
Can you specify the test scenario covered by this automation? 
Also are we going to cover negative scenario in a different PR?

File: test/secret/secret_s1_test.go [hunk 4]
```diff
@@ -181,6 +186,10 @@ var _ = Describe("Secret Test for SVA S1", func() {
 
 			// Verify Secrets on ServerConf on Pod
 			testenv.VerifySplunkServerConfSecrets(deployment, testenvInstance, verificationPods, secretStruct.Data, false)
+
+			// Verify Secrets via api access on Pod
+			secretStruct, err = testenv.GetSecretStruct(deployment, testenvInstance.GetName(), namespaceScopedSecretName)
+			testenv.VerifySplunkSecretViaAPI(deployment, testenvInstance, verificationPods, secretStruct.Data)
```
Comment (pdhanoyasplunk) [inline test/secret/secret_s1_test.go:192]: This looks like a positive match? However we need a negative verification here ?

File: test/testenv/util.go (hunk 3 above)
Comment (tpereirasplunk) [reply test/testenv/util.go:526]: plan to use the ExecPodCommand for multiple verification. So added a function for duplicate code

**Review by pdhanoyasplunk — APPROVED** [review]

**Review by smohan-splunk — APPROVED** [review]

**Pull request merged**